Optional background tracking taking 5 random times, instead of 5 regular times per day

DigitalSeneca opened the pull request:
fixes #9 

- Schedule a daily task at 00:00 that generated 5 random times for the day to collect the locations and send to the api

epou commented:
Important for iOS:
> registerPeriodicTask is only supported on iOS 13+, it might run for only 30 seconds due to iOS restrictions, but doesn't start immediately, rather iOS will schedule it as per user's App usage pattern.

epou commented on lib/utils/BackgroundTracking.dart:
```diff
@@ -0,0 +1,53 @@
+
+
+class BackgroundTracking{
+  static List<TimeOfDay> getRandomTimes() {
```
add argument `int num_samples` to make it more generic

epou commented on lib/utils/BackgroundTracking.dart:
```diff
@@ -0,0 +1,53 @@
+    return randomTimes;
+  }
+
+  static void fiveTimesPerDayTracking() {
```
use a more generic naming, such as: `scheduleMultipleTrackingTask`, which has an argument `int num_tasks`

epou commented on lib/utils/BackgroundTracking.dart:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  static void fiveTimesPerDayTracking() {
+    var randomTimes = getRandomTimes();
```
call getRandomTimes(num_samples=num_tasks)

epou commented on lib/utils/BackgroundTracking.dart:
```diff
@@ -0,0 +1,53 @@
+        'trackingTask',
+        initialDelay: Duration(hours: randomTimes[i].hour, minutes: randomTimes[i].minute),
+      );
+    }
```
```dart
randomTimes.asMap().forEach((index, time) {
    Workmanager().registerOneOffTask(
      'tracking_task_$index',
      'trackingTask',
      initialDelay: Duration(hours: time.hour, minutes: time.minute),
    );
  });
  ```

epou commented on lib/main.dart:
```diff
@@ -43,19 +54,12 @@ void callbackDispatcher() {
+        BackgroundTracking.trackingTask();
+        break;
+      case 'fiveTimesPerDayTracking':
+        BackgroundTracking.fiveTimesPerDayTracking();
```
scheduleMultipleTrackingTask(5)

epou commented on lib/utils/BackgroundTracking.dart:
```diff
@@ -0,0 +1,53 @@
+    var randomTimes = getRandomTimes();
+
+    for (var i = 0; i < 5; i++) {
+      Workmanager().registerOneOffTask(
```
ensure no task retry on failure, default is exponential: https://pub.dev/packages/workmanager#backoffpolicy

DigitalSeneca replied:
I can put `backoffPolicy: BackoffPolicy.linear`, and also the constraint to only execute with internet connection `constraints: Constraints(networkType: NetworkType.connected)`, but I am not seeing a way to not retry on failure. `registerOneOffTask` returns a void so it's not possible to now if it succeeded also

epou commented on lib/main.dart:
```diff
@@ -33,7 +34,17 @@ void main({String env = 'prod'}) async {
+    'fiveTimesPerDayTracking',
+    'fiveTimesPerDayTracking',
+    frequency: Duration(days: 1),
+    initialDelay: timeUntilMidnight,
```
add `tag: "backgroundTracking"`

epou commented:
The idea of using a unique tag is that when the user disables the background tracking in the settings page, pending tasks could be canceled by using: `Workmanager().cancelByTag("tag");`

DigitalSeneca commented:
> The idea of using a unique tag is that when the user disables the background tracking in the settings page, pending tasks could be canceled by using: Workmanager().cancelByTag("tag");

Another option is to just `await Workmanager().cancelAll();`

epou commented:
In the scenario where various tasks are concurrently running in the background, employing `Workmanager().cancelAll();` might not be the most efficient approach. Instead, it's advisable to selectively cancel only those tasks associated with backgroundTracking. That's why I came up with the solution of employing a consistent tag for all tracking-related tasks. This tagging system allows for grouping these tasks together, facilitating their cancellation using a unified tag, regardless of individual task names.

DigitalSeneca commented on ios/Runner/Info.plist:
```diff
@@ -0,0 +1,208 @@
+<dict>
+	<key>BGTaskSchedulerPermittedIdentifiers</key>
+	<array>
+        <string>task-identifier</string>
+        <string>be.tramckrijte.workmanagerExample.iOSBackgroundAppRefresh</string>
```
I added Info.plist, just adding those 2 lines for the iOS setup. So the merge conflict resolution with #125 should be easy

epou replied:
Theses two tasks does not exist in our code. Ours are named:
- scheduleDailyTasks
- trackingTask

See: https://github.com/fluttercommunity/flutter_workmanager/blob/d7e6ba5ef3796fafe54d9739e481f01058751b4d/example/ios/Runner/Info.plist#L5-L16

DigitalSeneca commented on ios/Runner/Info.plist:
```diff
@@ -0,0 +1,208 @@
+	<string>We need access to this feature in order to show results suited for you.</string>
+	<key>NSUserTrackingUsageDescription</key>
+	<string>We use your device identifier to track you across sessions and to save your reports accordingly.</string>
+	<key>UIBackgroundModes</key>
+	<array>
+		<string>fetch</string>
+		<string>location</string>
+		<string>processing</string>
```
Those are also needed for iOS but we already had them in there

DigitalSeneca commented:
> Important for iOS:
> 
> > registerPeriodicTask is only supported on iOS 13+, it might run for only 30 seconds due to iOS restrictions, but doesn't start immediately, rather iOS will schedule it as per user's App usage pattern.
> > ⚠️ On iOS 13+, adding a BGTaskSchedulerPermittedIdentifiers key to the Info.plist for new BGTaskScheduler API disables the performFetchWithCompletionHandler and setMinimumBackgroundFetchInterval methods, which means you cannot use both old Background Fetch and new registerPeriodicTask at the same time, you have to choose one based on your minimum iOS target version. For details see [Apple Docs](https://developer.apple.com/documentation/uikit/app_and_environment/scenes/preparing_your_ui_to_run_in_the_background/using_background_tasks_to_update_your_app)

I've added the setup for iOS but I can't test it without a device
I've also removed some transistorsoft settings that were still there from the previous location system

> In the scenario where various tasks are concurrently running in the background, employing Workmanager().cancelAll(); might not be the most efficient approach. Instead, it's advisable to selectively cancel only those tasks associated with backgroundTracking. That's why I came up with the solution of employing a consistent tag for all tracking-related tasks. This tagging system allows for grouping these tasks together, facilitating their cancellation using a unified tag, regardless of individual task names.

Okey, so the daily task is still executed, but if there's no permission enabled, we just skip it and don't schedule the 5 tasks for the day?

epou commented on ios/Runner/Info.plist:
```diff
@@ -0,0 +1,208 @@
+	<key>CFBundleExecutable</key>
+	<string>$(EXECUTABLE_NAME)</string>
+	<key>CFBundleIdentifier</key>
+	<string>cat.ibeji.tigatrapp</string>
```
Add:

```xml
<key>MinimumOSVersion</key>
<string>13.0</string>
```

epou commented:
> Okey, so the daily task is still executed, but if there's no permission enabled, we just skip it and don't schedule the 5 tasks for the day?

I was considering halting all background tracking tasks if the user disables that option in the settings page. However, if you believe it's more efficient to keep the scheduler running and simply refrain from scheduling any tasks when the option is disabled, feel free to proceed with that approach.

DigitalSeneca commented:
> > Okey, so the daily task is still executed, but if there's no permission enabled, we just skip it and don't schedule the 5 tasks for the day?
> 
> I was considering halting all background tracking tasks if the user disables that option in the settings page. However, if you believe it's more efficient to keep the scheduler running and simply refrain from scheduling any tasks when the option is disabled, feel free to proceed with that approach.

We can halt all tracking tasks, including the one with tag 'scheduleDailyTasks', but then I don't see the performance improvement of cancelling by tag instead of `cancelAll();`, when we're going to cancel all tasks anyways

epou commented:
Okay! We should only need extra care if in the future we add any other kind of task that is not related to background tracking, and only cancel these.

DigitalSeneca commented on lib/main.dart:
```diff
@@ -43,19 +53,12 @@ void callbackDispatcher() {
+        await BackgroundTracking.scheduleMultipleTrackingTask(5);
+        break;
     }
     return Future.value(true);
```
@epou With this line, the workmanager should not do the automatic retry on Android, so no need to actually cancel the task if it fails https://pub.dev/packages/workmanager#work-result

epou replied:
Awesome! So then, `backoffPolicy` should be removed

epou commented on lib/main.dart:
```diff
@@ -33,7 +31,19 @@ void main({String env = 'prod'}) async {
+    tag: 'scheduleDailyTasks',
+    frequency: Duration(days: 1),
+    initialDelay: timeUntilMidnight,
+    backoffPolicy: BackoffPolicy.linear,
```
remove

epou commented on ios/Runner/AppDelegate.swift:
```diff
@@ -16,6 +17,11 @@ import Firebase
+    WorkmanagerPlugin.registerBGProcessingTask(withIdentifier: "trackingTask")
+    WorkmanagerPlugin.registerBGProcessingTask(withIdentifier: "scheduleDailyTasks")
+    // Register a periodic task in iOS 13+
+    WorkmanagerPlugin.registerPeriodicTask(withIdentifier: "be.tramckrijte.workmanagerExample.iOSBackgroundAppRefresh", frequency: NSNumber(value: 86400))
```
i guess this is not required... it belongs to the example here https://github.com/fluttercommunity/flutter_workmanager/blob/d7e6ba5ef3796fafe54d9739e481f01058751b4d/example/ios/Runner/AppDelegate.swift#L23

epou commented on lib/utils/BackgroundTracking.dart:
```diff
@@ -0,0 +1,63 @@
+        initialDelay: Duration(hours: time.hour, minutes: time.minute),
+        tag: 'trackingTask',
+        constraints: Constraints(networkType: NetworkType.connected),
+        backoffPolicy: BackoffPolicy.linear,
```
remove backoffpolicy

epou commented on ios/Runner/AppDelegate.swift:
```diff
@@ -16,6 +17,9 @@ import Firebase
     GeneratedPluginRegistrant.register(with: self)
+    UIApplication.shared.setMinimumBackgroundFetchInterval(TimeInterval(60*15))
+    WorkmanagerPlugin.registerBGProcessingTask(withIdentifier: "trackingTask")
+    WorkmanagerPlugin.registerBGProcessingTask(withIdentifier: "scheduleDailyTasks")
```
Not sure but... Shouldn't this be
`WorkmanagerPlugin.registerPeriodicTask(withIdentifier: "scheduleDailyTasks", frequency: NSNumber(value: 24 * 60 * 60))
`

epou commented on ios/Runner/Info.plist:
```diff
@@ -0,0 +1,210 @@
+<dict>
+	<key>BGTaskSchedulerPermittedIdentifiers</key>
+	<array>
+		<string>com.example.mosquito_alert_app.scheduleDailyTasks</string>
+    	<string>com.example.mosquito_alert_app.trackingTask</string>
```
I guess both value must be equal to the ones declared in AppDelegate.swift, so then:
```
    <string>trackingTask</string>
    <string>scheduleDailyTasks</string>
```
